cmdstanpy/cmdstan_args.py

```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
-                    self.metric = 'diag_e'
+                    self.metric_type = 'diag_e'
                 elif self.metric in ['dense', 'dense_e']:
-                    self.metric = 'dense_e'
+                    self.metric_type = 'dense_e'
                 elif self.metric in ['unit', 'unit_e']:
-                    self.metric = 'unit_e'
+                    self.metric_type = 'unit_e'
                 else:
                     if not os.path.exists(self.metric):
                         raise ValueError('no such file {}'.format(self.metric))
                     dims = read_metric(self.metric)
+                    if len(dims) == 1:
+                        self.metric_type = 'diag_e'
+                    else:
+                        self.metric_type = 'dense_e'
+                    self.metric_file = self.metric
+            elif isinstance(self.metric, Dict):
+                if 'inv_metric' not in self.metric:
+                    raise ValueError(
+                        'Entry "inv_metric" not found in metric dict.'
+                    )
+                dims = list(np.asarray(self.metric['inv_metric']).shape)
+                if len(dims) == 1:
+                    self.metric_type = 'diag_e'
+                else:
+                    self.metric_type = 'dense_e'
+                dict_file = create_named_text_file(
+                    dir=_TMPDIR, prefix="metric", suffix=".json"
+                )
+                write_stan_json(dict_file, self.metric)
+                self.metric_file = dict_file
             elif isinstance(self.metric, (list, tuple)):
                 if len(self.metric) != chains:
                     raise ValueError(
-                        'number of metric files must match number of chains,'
-                        ' found {} metric files for {} chains'.format(
+                        'Number of metric files must match number of chains,'
+                        ' found {} metric files for {} chains.'.format(
                             len(self.metric), chains
                         )
                     )
-                names_set = set(self.metric)
-                if len(names_set) != len(self.metric):
-                    raise ValueError(
-                        'each chain must have its own metric file,'
-                        ' found duplicates in metric files list.'
-                    )
-                for i, metric in enumerate(self.metric):
-                    if not os.path.exists(metric):
-                        raise ValueError('no such file {}'.format(metric))
-                    if i == 0:
-                        dims = read_metric(metric)
-                    else:
-                        dims2 = read_metric(metric)
-                        if len(dims) != len(dims2):
+                if all(isinstance(elem, Dict) for elem in self.metric):
+                    metric_files: List[str] = []
+                    for i, metric in enumerate(self.metric):
```

Comment:
If we move this loop up one block and check dict / str for each element, we could then also accept user to give a combination of dict and str

Reply:
we could, but should we?  I don't see a good use case.

Reply:
True, probably causes more damage :D